### DISCUSSION TITLE
How to get a copy of Material or a deep copy of MaterialInstance

### DISCUSSION THREAD

**Discussion by RockVampire** [google/filament#4585] [Q&A]
I'm adding the feature of picking entity in filament on Android . First I solved the problem of finding the entity with bullet's raycast as you said in #623.
Now, I'm trying to show the target object by changing their `Material`. 
I find the way in #4109 . But a MaterialInstance may be used by not only one entity in a scene , and the other entity be affected concurrently. 
I also tried the way in #3992 like this 

```
						var renderableManager = modelViewer.engine.renderableManager
						otherMaterialInstance =renderableManager.getMaterialInstanceAt(key,0).material.createInstance()
						otherMaterialInstance.setParameter("baseColorFactor",Colors.RgbaType.SRGB,0f,1f,0f,0.5f)
						renderableManager.setMaterialInstanceAt(key,0,otherMaterialInstance)
```
It doesn't work and I got crash occasionally.
I think maybe the problem is `Material ` shared by two `MaterialInstance`?
I know I can get a New Material Instance by a matc file . But the effect we want is something based on the original Material such as high lighting or just changing the parameter "metallic". So ,if that's the problem , how can I get a `Material`  by the original Material I got from the gltf file.
Or there's another mistake?

**Comment by romainguy** [comment]
What is the crash? A `Material` can have many `MaterialInstance`. It's the whole point of material instances. Think of a `MaterialInstance` as a set of parameter values for a `Material`, nothing more.

**Reply by RockVampire** [reply]
I don't add source code dependency in my testing project so I can trace the crash deeper.Only got the crash log 
` Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x1400000013 `
like every other native crash on Android.Something I know is just the crash always comes after the code in my question run 3-4 times.
So ,as you saying ,my code is correct and the `Material` will use the parameters in the new `MaterialInstance` when I called `RenderableManager.setMaterialInstanceAt`?
Without the crash , I just want to change `Material` of one `Renderable`. But it seems there's some `Renderable` sharing the `Material`.So what can I do to make the modification effect only one `Renderable`.Should I get a new `Material`? and How ?

**Reply by romainguy** [reply]
It doesn't matter if multiple renderables are sharing the material. You can give different material instances to different renderables.

**Reply by RockVampire** [reply]
Okay, I think I've got what you said. 
And I got the reason of the crash , I select the 'physic ground entity' ,and trying to alter its material however it doesn't exist.
Emmm, but I meet another problem. When I'm trying to use ` MaterialInstance` I got with `RenderableManager.getMaterialInstanceAt(key,0).material.createInstance()`, the renderable entity becomes all black.It seems I've got an empty ` MaterialInstance` by `createInstance()`. It will work unless I set all the parameters manually. Is it correct?

And according to the suppose，I also tried to get the new `MaterialInstance`  like this `otherMaterialInstance = MaterialInstance.duplicate(oldMaterialInstance,"light")`.Then I got a new crash when calling `RenderableManager.setMaterialInstanceAt(it,0,otherMaterialInstance)`
The crash Log remind me I'm calling method on destroyed Material.

```
 E/InputEventReceiver: Exception dispatching input event.
 E/MessageQueue-JNI: Exception in MessageQueue callback: handleReceiveCallback
E/MessageQueue-JNI: java.lang.IllegalStateException: Calling method on destroyed Material
```
I debug into the method and found it crashed in the row ` int required = materialInstance.getMaterial().getRequiredAttributesAsInt();` when 'getRequiredAttributesAsInt()'
I compared the two `MaterialInstance` , it seems the `MaterialInstance` I got from MaterialInstance.duplicate() comes from the constructor   `MaterialInstance(@NonNull Material material, long nativeMaterialInstance)` . It seems missing ‘mNativeMaterial = nGetMaterial(mNativeObject);’ by contract. 
Is it the reason of the crash ? and is it a bug?
